### PR TITLE
Support CREATE FUNCTION execution

### PR DESCRIPTION
**Pull request by caithagoras**
```
== RELEASE NOTES ==

General Changes
* ...
* ...
```

### CODE REVIEW

**Comment by caithagoras** [comment]
Ready for initial review. Will add more tests for commit `Support MySQL based function namespace manager`

File: presto-main/src/main/java/com/facebook/presto/metadata/FunctionManager.java [hunk 1]
```diff
@@ -169,23 +171,29 @@ public void registerBuiltInFunctions(List<? extends BuiltInFunction> functions)
...
-
-    private FunctionHandle resolveFunction(Optional<Session> session, QualifiedName name, List<TypeSignatureProvider> parameterTypes)
     {
         QualifiedFunctionName functionName;
         if (!name.getPrefix().isPresent()) {
             functionName = QualifiedFunctionName.of(DEFAULT_NAMESPACE, name.getSuffix());
         }
         else {
-            functionName = QualifiedFunctionName.of(name.getOriginalParts());
+            checkArgument(name.getOriginalParts().size() == 3, "Require 3 parts for qualified function name");
```
Comment (rongrong) [inline presto-main/src/main/java/com/facebook/presto/metadata/FunctionManager.java]: This should throw a `PrestoException` with proper error message that's easy to understand for user. Something along the line of `Reference to non-builtin functions must be using qualified function name in the form of "a.c.b"` if `originalParts = 2`, and throwing `FUNCTION_NOT_FOUND` is reasonable if `originalParts > 3`. The 3 parts function name should be a side effect of how function namespaces are managed. Since all sql function names would have 3 parts, a >3 parts function can not be resolved.

File: presto-main/src/main/java/com/facebook/presto/metadata/FunctionManager.java [hunk 2]
```diff
@@ -219,13 +227,15 @@ private FunctionHandle resolveFunction(Optional<Session> session, QualifiedName
...
-        throw new PrestoException(FUNCTION_NOT_FOUND, constructFunctionNotFoundErrorMessage(name.toString(), parameterTypes, candidates));
+        throw new PrestoException(FUNCTION_NOT_FOUND, constructFunctionNotFoundErrorMessage(functionName.toString(), parameterTypes, candidates));
     }
 
     @Override
     public FunctionMetadata getFunctionMetadata(FunctionHandle functionHandle)
     {
-        return functionNamespaces.get(functionHandle.getFunctionNamespace()).getFunctionMetadata(functionHandle);
+        Optional<FunctionNamespaceManager<?>> functionNamespaceManager = getServingFunctionNamespaceManager(functionHandle.getFunctionNamespace());
+        checkArgument(functionNamespaceManager.isPresent(), "Cannot find function namespace for '%s'", functionHandle.getFunctionNamespace());
```
Comment (rongrong) [inline presto-main/src/main/java/com/facebook/presto/metadata/FunctionManager.java]: We already have a function handle here, how could there be no FunctionNamespaceManager? Why do we need this new logic?

File: presto-spi/src/main/java/com/facebook/presto/spi/function/SqlInvokedFunction.java [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.spi.function;
+
+import java.util.List;
+import java.util.Optional;
+
+public interface SqlInvokedFunction
```
Comment (rongrong) [inline presto-spi/src/main/java/com/facebook/presto/spi/function/SqlInvokedFunction.java]: What are the other potential kind of `SqlInvokedFunction` other than `SqlInvokedRegularFunction`?

File: presto-spi/src/main/java/com/facebook/presto/spi/function/SqlFunctionId.java [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.spi.function;
+
+public interface SqlFunctionId
+{
+    QualifiedFunctionName getFunctionName();
+
+    String getValue();
```
Comment (rongrong) [inline presto-spi/src/main/java/com/facebook/presto/spi/function/SqlFunctionId.java]: This is vague, what does "value" mean here?

File: presto-main/src/test/java/com/facebook/presto/metadata/TestFunctionManager.java [hunk 5]
```diff
@@ -79,7 +65,7 @@ public void testIdentityCast()
     public void testExactMatchBeforeCoercion()
     {
         TypeRegistry typeManager = new TypeRegistry();
-        FunctionManager functionManager = createFunctionManager(typeManager);
+        FunctionManager functionManager = new FunctionManager(typeManager, new BlockEncodingManager(typeManager), new FeaturesConfig());
```
Comment (rongrong) [inline presto-main/src/test/java/com/facebook/presto/metadata/TestFunctionManager.java]: What is this change about?

File: presto-function-namespace-managers/src/main/java/com/facebook/presto/sqlfunction/db/FunctionNamespaceDao.java [hunk 6]
```diff
@@ -0,0 +1,176 @@
...
+            "  `id` bigint(20) NOT NULL AUTO_INCREMENT,\n" +
+            "  `function_id` varchar(700) NOT NULL,\n" +
+            "  `version` bigint(20) unsigned NOT NULL,\n" +
+            "  `function_name` varchar(512) NOT NULL,\n" +
+            "  `parameters` varchar(2048) NOT NULL,\n" +
+            "  `return_type` varchar(128) NOT NULL,\n" +
+            "  `language` varchar(64) NOT NULL,\n" +
+            "  `deterministic` tinyint(1) NOT NULL,\n" +
+            "  `called_on_null_input` tinyint(1) NOT NULL,\n" +
+            "  `body` mediumtext,\n" +
```
Comment (rongrong) [inline presto-function-namespace-managers/src/main/java/com/facebook/presto/sqlfunction/db/FunctionNamespaceDao.java]: I think text is enough. Not sure in what situation we would want function implementations to be > 64KB.

File: presto-function-namespace-managers/src/main/java/com/facebook/presto/sqlfunction/db/FunctionNamespaceDao.java [hunk 7]
```diff
@@ -0,0 +1,176 @@
...
+            "  `description` text,\n" +
+            "  `created_time` TIMESTAMP NOT NULL DEFAULT CURRENT_TIMESTAMP,\n" +
+            "  `modified_time` TIMESTAMP NOT NULL DEFAULT CURRENT_TIMESTAMP ON UPDATE CURRENT_TIMESTAMP,\n" +
+            "  PRIMARY KEY (`id`),\n" +
+            "  UNIQUE KEY `function_id_version` (`function_id`,`version`),\n" +
```
Comment (rongrong) [inline presto-function-namespace-managers/src/main/java/com/facebook/presto/sqlfunction/db/FunctionNamespaceDao.java]: Semantically `function_id` should be a function name + parameters types. However this is not enforced on the tables. So if function_id does not have the same uniqueness as function name + parameters this implementation is in trouble.

File: presto-function-namespace-managers/src/main/java/com/facebook/presto/sqlfunction/db/FunctionNamespaceDao.java [hunk 8]
```diff
@@ -0,0 +1,176 @@
...
+        @RegisterArgumentFactory(SqlParametersArgumentFactory.class),
+        @RegisterArgumentFactory(TypeSignatureArgumentFactory.class),
+})
+public interface FunctionNamespaceDao
+{
+    @SqlUpdate("CREATE TABLE `<table_name>` (\n" +
+            "  `id` bigint(20) NOT NULL AUTO_INCREMENT,\n" +
+            "  `function_id` varchar(700) NOT NULL,\n" +
+            "  `version` bigint(20) unsigned NOT NULL,\n" +
+            "  `function_name` varchar(512) NOT NULL,\n" +
```
Comment (rongrong) [inline presto-function-namespace-managers/src/main/java/com/facebook/presto/sqlfunction/db/FunctionNamespaceDao.java]: a 512 character function name sounds way too long. Seems like this would include the function namespace as well?

File: presto-function-namespace-managers/src/main/java/com/facebook/presto/sqlfunction/db/FunctionNamespaceDao.java [hunk 9]
```diff
@@ -0,0 +1,176 @@
...
+            "    SELECT\n" +
+            "        `function_id`,\n" +
+            "        MAX(`version`) `version`\n" +
+            "    FROM `<table_name>`\n" +
+            "    GROUP BY\n" +
+            "        `function_id`\n" +
+            ") v\n" +
+            "    ON t.`function_id` = v.`function_id`\n " +
+            "    AND t.`version` = v.`version`")
+    List<SqlInvokedFunction> listFunctions(@Define("table_name") String tableName);
```
Comment (rongrong) [inline presto-function-namespace-managers/src/main/java/com/facebook/presto/sqlfunction/db/FunctionNamespaceDao.java]: I think `listFunctions` only need to return a `SqlFunction`. This is used for `SHOW FUNCTIONS`, which only needs the metadata, not the body, or the version. If you cannot find a good class, maybe just return `List<FunctionMetadata>`.

**Comment by caithagoras** [comment]
@rongrong Thanks for the review! `Support MySQL based function namespace manager` is the last commit. The commits beyond have not been finished and will be moved to a separate PR for the purpose of executing CREATE FUNCTION.

File: presto-main/src/test/java/com/facebook/presto/metadata/TestFunctionManager.java (hunk 5 above)
Comment (caithagoras) [reply presto-main/src/test/java/com/facebook/presto/metadata/TestFunctionManager.java]: This belongs to a different PR.

File: presto-function-namespace-managers/src/main/java/com/facebook/presto/sqlfunction/db/FunctionNamespaceDao.java (hunk 9 above)
Comment (caithagoras) [reply presto-function-namespace-managers/src/main/java/com/facebook/presto/sqlfunction/db/FunctionNamespaceDao.java]: This `listFunctions` is at the end of the call stack for `Metadata.listFunction()`. We'll need refactor
a few layers of methods in order to do that, including `Metadata.listFunction()`, `FunctionManager.listFunction()`, `FunctionNamespaceManager.listFunction()`. For now, I'm adding a TODO in `FunctionNamespaceManager.listFunction()`

File: presto-spi/src/main/java/com/facebook/presto/spi/function/CatalogSchemaPrefix.java [hunk 10]
```diff
@@ -0,0 +1,90 @@
...
+import java.util.Optional;
+
+import static java.util.Objects.requireNonNull;
+
+public class CatalogSchemaPrefix
+{
+    private final String catalogName;
+    private final Optional<String> schemaName;
+
+    public CatalogSchemaPrefix(String catalogName, Optional<String> schemaName)
```
[the remark this answers is not in the document]
Comment (caithagoras) [reply presto-spi/src/main/java/com/facebook/presto/spi/function/CatalogSchemaPrefix.java]: This class is `CatalogSchemaPrefix`. It is used to represent `catalog` or `catalog.schema`. The current usage in this commit is represent the list of "serving namespaces" of a function namespace manager. Example:
```
serving-namespaces=testing,dev.di
```
A "serving namespaces" is basically `List<CatalogSchemaPrefix>`

File: presto-spi/src/main/java/com/facebook/presto/spi/function/QualifiedFunctionName.java [hunk 11]
```diff
@@ -13,168 +13,75 @@
...
-    public List<String> getOriginalParts()
+    public static QualifiedFunctionName of(CatalogSchemaName functionNamespace, String name)
     {
-        return originalParts;
+        return new QualifiedFunctionName(functionNamespace, name);
     }
 
-    public String getSuffix()
+    // TODO: Examine all callers to limit the usage of the method
+    public String getUnqualifiedName()
```
[the remark this answers is not in the document]
Comment (caithagoras) [reply presto-spi/src/main/java/com/facebook/presto/spi/function/QualifiedFunctionName.java]: I do think `suffix` is accurate in certain cases. For example, it makes sense to say `c` is the suffix of a String `a.b.c`. It also make sense to say `c` is the suffix of a `QualifiedName("a.b.c")`  where it is consisted of `List<String> parts`.

However, we're explicitly defining a `QualifiedFunctionName` in terms of `(CatalogSchemaName functionNamespace, String unqualifiedName)`. Also, using this name indicates an issue. Eventually, we'll be able to remove this method completedly.

File: presto-function-namespace-managers/src/main/java/com/facebook/presto/sqlfunction/db/FunctionNamespaceDao.java [hunk 12]
```diff
@@ -0,0 +1,176 @@
...
+        @RegisterArgumentFactory(QualifiedFunctionNameArgumentFactory.class),
+        @RegisterArgumentFactory(SqlFunctionIdArgumentFactory.class),
+        @RegisterArgumentFactory(SqlParametersArgumentFactory.class),
+        @RegisterArgumentFactory(TypeSignatureArgumentFactory.class),
+})
+public interface FunctionNamespaceDao
+{
+    @SqlUpdate("CREATE TABLE `<table_name>` (\n" +
+            "  `id` bigint(20) NOT NULL AUTO_INCREMENT,\n" +
+            "  `function_id` varchar(700) NOT NULL,\n" +
```
[the remark this answers is not in the document]
Comment (caithagoras) [reply presto-function-namespace-managers/src/main/java/com/facebook/presto/sqlfunction/db/FunctionNamespaceDao.java]: This is because the max key length for MySQL is 768.

File: presto-function-namespace-managers/src/main/java/com/facebook/presto/sqlfunction/db/FunctionNamespaceDao.java (hunk 7 above)
Comment (caithagoras) [reply presto-function-namespace-managers/src/main/java/com/facebook/presto/sqlfunction/db/FunctionNamespaceDao.java]: Removed for now.

File: presto-function-namespace-managers/src/main/java/com/facebook/presto/sqlfunction/db/FunctionNamespaceDao.java (hunk 6 above)
Comment (caithagoras) [reply presto-function-namespace-managers/src/main/java/com/facebook/presto/sqlfunction/db/FunctionNamespaceDao.java]: Removed for now.

File: presto-function-namespace-managers/src/main/java/com/facebook/presto/sqlfunction/db/FunctionNamespaceDao.java (hunk 8 above)
Comment (caithagoras) [reply presto-function-namespace-managers/src/main/java/com/facebook/presto/sqlfunction/db/FunctionNamespaceDao.java]: Yes, it includes the function namespace.

Commit removed for now.

File: presto-main/src/main/java/com/facebook/presto/metadata/FunctionManager.java (hunk 2 above)
Comment (caithagoras) [reply presto-main/src/main/java/com/facebook/presto/metadata/FunctionManager.java]: The `functionNamespaces` variables stores the mapping from prefix to manager, and it is acceptable that the prefix only contains the catalog.

File: presto-spi/src/main/java/com/facebook/presto/spi/function/SqlFunctionId.java (hunk 4 above)
Comment (caithagoras) [reply presto-spi/src/main/java/com/facebook/presto/spi/function/SqlFunctionId.java]: Renamed to `getId()`. This method effectively returns `toString`, the String representation of the `FunctionId`.

File: presto-spi/src/main/java/com/facebook/presto/spi/function/SqlInvokedFunction.java (hunk 3 above)
Comment (caithagoras) [reply presto-spi/src/main/java/com/facebook/presto/spi/function/SqlInvokedFunction.java]: Convert `SqlInvokedFunction`, `SqlFunctionId`, and `SqlFunctionHandle` to concrete classes.

File: presto-spi/src/main/java/com/facebook/presto/spi/function/FunctionNamespaceManager.java [hunk 13]
```diff
@@ -30,15 +30,15 @@
...
+     * {@link #abort(FunctionNamespaceTransactionHandle)} is called.
      */
     void commit(FunctionNamespaceTransactionHandle transactionHandle);
 
     /**
      * Rollback the transaction. Will be called at most once and will not be called if
      * {@link #commit(FunctionNamespaceTransactionHandle)} is called.
      */
-    void rollback(FunctionNamespaceTransactionHandle transactionHandle);
+    void abort(FunctionNamespaceTransactionHandle transactionHandle);
```
Comment (rongrong) [inline presto-spi/src/main/java/com/facebook/presto/spi/function/FunctionNamespaceManager.java]: Should this be a different commit?

Comment (caithagoras) [reply presto-spi/src/main/java/com/facebook/presto/spi/function/FunctionNamespaceManager.java]: Done

File: presto-main/src/main/java/com/facebook/presto/metadata/FunctionManager.java [hunk 14]
```diff
@@ -331,20 +343,20 @@ private FunctionHandle lookupFunction(
...
+        CatalogSchemaPrefix bestMatch = null;
         FunctionNamespaceManager<?> servingFunctionNamespaceManager = null;
 
-        for (Map.Entry<QualifiedFunctionName.Prefix, FunctionNamespaceManager<?>> functionNamespace : functionNamespaces.entrySet()) {
-            if (functionNamespace.getKey().contains(functionPrefix) && (bestMatchNamespace == null || bestMatchNamespace.contains(functionNamespace.getKey()))) {
-                bestMatchNamespace = functionNamespace.getKey();
-                servingFunctionNamespaceManager = functionNamespace.getValue();
+        for (Map.Entry<CatalogSchemaPrefix, FunctionNamespaceManager<?>> entry : functionNamespaces.entrySet()) {
+            CatalogSchemaPrefix prefix = entry.getKey();
+            if (prefix.includes(functionNamespace) && (bestMatch == null || bestMatch.includes(prefix))) {
```
Comment (rongrong) [inline presto-main/src/main/java/com/facebook/presto/metadata/FunctionManager.java]: Maybe have a `includesButNotEqual` instead. Currently this is relying on same prefix would not appear more than once.

File: presto-main/src/main/java/com/facebook/presto/sql/gen/FunctionCallCodeGenerator.java [hunk 15]
```diff
@@ -49,7 +49,7 @@ public BytecodeNode generateCall(FunctionHandle functionHandle, BytecodeGenerato
         }
 
         return context.generateCall(
-                functionManager.getFunctionMetadata(functionHandle).getName().getSuffix(),
+                functionManager.getFunctionMetadata(functionHandle).getName().getUnqualifiedName(),
```
Comment (rongrong) [inline presto-main/src/main/java/com/facebook/presto/sql/gen/FunctionCallCodeGenerator.java]: I think I mentioned before that `getSuffix` is more accurate than `getUnqualifiedName`. If you prefer to use "unqualified name", it's more appropriate to have a utility function `unqualifyName(qualifiedName)`. I'm not sure whether you can call the last part of a qualified name "unqualified name".

File: presto-spi/src/main/java/com/facebook/presto/spi/function/QualifiedFunctionName.java (hunk 11 above)
[the remark this answers is not in the document]
Comment (rongrong) [reply presto-spi/src/main/java/com/facebook/presto/spi/function/QualifiedFunctionName.java]: That should be defined as `(CatalogSchemaName functionNamespace, String suffix)` then.

File: presto-spi/src/main/java/com/facebook/presto/spi/function/SqlFunctionHandle.java [hunk 16]
```diff
@@ -11,49 +11,35 @@
...
 import com.fasterxml.jackson.annotation.JsonProperty;
 
-import java.util.List;
 import java.util.Objects;
 
 import static java.util.Objects.requireNonNull;
-import static java.util.stream.Collectors.joining;
 
-public class SqlInvokedRegularFunctionHandle
+public class SqlFunctionHandle
```
Comment (rongrong) [inline presto-spi/src/main/java/com/facebook/presto/spi/function/SqlFunctionHandle.java]: Add `@Experimental` to all classes introduced to spi.

File: presto-sql-function/src/main/java/com/facebook/presto/sqlfunction/AbstractSqlInvokedFunctionNamespaceManager.java [hunk 17]
```diff
@@ -127,6 +127,11 @@ public final FunctionMetadata getFunctionMetadata(FunctionHandle functionHandle)
         return metadataByHandle.getUnchecked((SqlFunctionHandle) functionHandle);
     }
 
+    protected void refreshFunctionsCache(QualifiedFunctionName functionName)
```
Comment (rongrong) [inline presto-sql-function/src/main/java/com/facebook/presto/sqlfunction/AbstractSqlInvokedFunctionNamespaceManager.java]: Where is this used?

Comment (caithagoras) [reply presto-sql-function/src/main/java/com/facebook/presto/sqlfunction/AbstractSqlInvokedFunctionNamespaceManager.java]: It is used by subclasses in which `createFunction` is implemented. This method can be called to update the functions cache on a new function or a new version fo a function is created.

File: presto-main/src/main/java/com/facebook/presto/sql/gen/FunctionCallCodeGenerator.java (hunk 15 above)
Comment (caithagoras) [reply presto-main/src/main/java/com/facebook/presto/sql/gen/FunctionCallCodeGenerator.java]: It is a convention to called the last part of a "fully qualified object" (file path, class name, etc) a unqualified name. Here are some reference:
1. https://stackoverflow.com/a/27966682
2. https://en.wikipedia.org/wiki/Fully_qualified_name

File: presto-main/src/main/java/com/facebook/presto/metadata/FunctionManager.java (hunk 14 above)
Comment (caithagoras) [reply presto-main/src/main/java/com/facebook/presto/metadata/FunctionManager.java]: I'm not sure what will be the behavior of the `includesButNotEqual` method you mentioned here.

File: presto-main/src/main/java/com/facebook/presto/execution/CreateFunctionTask.java [hunk 18]
```diff
@@ -0,0 +1,104 @@
...
+    @Override
+    public ListenableFuture<?> execute(CreateFunction statement, TransactionManager transactionManager, Metadata metadata, AccessControl accessControl, QueryStateMachine stateMachine, List<Expression> parameters)
+    {
+        Analyzer analyzer = new Analyzer(stateMachine.getSession(), metadata, sqlParser, accessControl, Optional.empty(), parameters, stateMachine.getWarningCollector());
+        analyzer.analyze(statement);
+        metadata.getFunctionManager().createFunction(createFunctionToSqlInvokedFunction(statement), statement.isReplace());
+        return immediateFuture(null);
+    }
+
+    private SqlInvokedFunction createFunctionToSqlInvokedFunction(CreateFunction statement)
```
Comment (rongrong) [inline presto-main/src/main/java/com/facebook/presto/execution/CreateFunctionTask.java]: the function name is a little strange. Maybe `createSqlInvokedFunction`?

File: presto-main/src/main/java/com/facebook/presto/execution/CreateFunctionTask.java [hunk 19]
```diff
@@ -0,0 +1,104 @@
...
+    {
+        if (statement.getFunctionName().getParts().size() != 3) {
+            throw new PrestoException(GENERIC_INTERNAL_ERROR, format("Invalid SqlFunction name: %s, require exactly 3 parts", statement.getFunctionName().getParts().size()));
```
Comment (rongrong) [inline presto-main/src/main/java/com/facebook/presto/execution/CreateFunctionTask.java]: Should this be an user error?

File: presto-main/src/main/java/com/facebook/presto/metadata/FunctionManager.java [hunk 20]
```diff
@@ -343,6 +347,15 @@ private FunctionHandle lookupFunction(
         throw new PrestoException(FUNCTION_NOT_FOUND, constructFunctionNotFoundErrorMessage(functionName, parameterTypes, candidates));
     }
 
+    private FunctionNamespaceManager<?> getRequiredServingFunctionNamespaceManager(CatalogSchemaName functionNamespace)
+    {
+        Optional<FunctionNamespaceManager<?>> functionNamespaceManager = getServingFunctionNamespaceManager(functionNamespace);
+        if (!functionNamespaceManager.isPresent()) {
+            throw new PrestoException(FUNCTION_NOT_FOUND, format("Cannot find function namespace '%s'", functionNamespace));
```
Comment (rongrong) [inline presto-main/src/main/java/com/facebook/presto/metadata/FunctionManager.java]: If this is called during `CREATE FUNCTION` then this error code is not appropriate. Maybe this class should just throw run time error and let caller decide what error code it should use.

File: presto-main/src/main/java/com/facebook/presto/metadata/FunctionManager.java (hunk 14 above)
Comment (rongrong) [reply presto-main/src/main/java/com/facebook/presto/metadata/FunctionManager.java]: the `includes` method in prefix returns true of it has a bigger or equal scope. I'm suggesting a method that would just return true if it actually has a bigger scope. For example `"a.b" includes "a.b" => true`, `"a.b" includes but not equal "a.b" => false`.

File: presto-sql-function/src/main/java/com/facebook/presto/sqlfunction/AbstractSqlInvokedFunctionNamespaceManager.java (hunk 17 above)
Comment (caithagoras) [reply presto-sql-function/src/main/java/com/facebook/presto/sqlfunction/AbstractSqlInvokedFunctionNamespaceManager.java]: Yes, it is in `presto-facebook` now that we have remove the plugin commit from the PR.

File: presto-main/src/main/java/com/facebook/presto/sql/gen/FunctionCallCodeGenerator.java (hunk 15 above)
Comment (rongrong) [reply presto-main/src/main/java/com/facebook/presto/sql/gen/FunctionCallCodeGenerator.java]: Again, those reference only said a name without prefix is a unqualified name, rather than saying the last part of a qualified name is a "unqualified name". The reference you mentioned also used "suffix", which I still think is better than "unqualified name". If you want to be stubborn on this I won't insist. But I think "suffix" is better than "unqualified name" in this case.

File: presto-main/src/main/java/com/facebook/presto/metadata/FunctionManager.java (hunk 20 above)
Comment (caithagoras) [reply presto-main/src/main/java/com/facebook/presto/metadata/FunctionManager.java]: Since different callers need to handle them differently, I'm removing this method and handle the cases of `!functionNamespaceManager.isPresent()` in each caller (2) separately.

File: presto-main/src/main/java/com/facebook/presto/sql/gen/FunctionCallCodeGenerator.java (hunk 15 above)
Comment (caithagoras) [reply presto-main/src/main/java/com/facebook/presto/sql/gen/FunctionCallCodeGenerator.java]: Looking at some of our existing classes:
- `QualifiedObjectName` is defined as (catalogName, schemaName, objectName)
- `QualifiedTablePrefix` is defined as (catalogName, schemaName, tableName)

What about defining `QualifiedFunctionName` is (functionNamespace, functionName)?

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java [hunk 21]
```diff
@@ -538,6 +540,26 @@ protected Scope visitCreateView(CreateView node, Optional<Scope> scope)
...
+
+            Type returnType = metadata.getType(parseTypeSignature(node.getReturnType()));
+            List<Field> fields = node.getParameters().stream()
+                    .map(parameter -> Field.newUnqualified(parameter.getName().getValue(), metadata.getType(parseTypeSignature(parameter.getType()))))
+                    .collect(toImmutableList());
+            Scope functionScope = Scope.builder()
+                    .withRelationType(RelationId.anonymous(), new RelationType(fields))
+                    .build();
+            Type bodyType = analyzeExpression(node.getBody(), functionScope).getExpressionTypes().get(NodeRef.of(node.getBody()));
+            metadata.getTypeManager().canCoerce(bodyType, returnType);
```
Comment (rongrong) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java]: I'm not sure whether there's implications at run time if the return type of body is not exactly the same as the given `returnType`. Is there valid reasons that we want to allow coercion? Why not just require the return type to be exact?

File: presto-main/src/main/java/com/facebook/presto/execution/CreateFunctionTask.java (hunk 18 above)
Comment (rongrong) [reply presto-main/src/main/java/com/facebook/presto/execution/CreateFunctionTask.java]: This comment was not addressed.

File: presto-main/src/main/java/com/facebook/presto/execution/CreateFunctionTask.java [hunk 22]
```diff
@@ -0,0 +1,104 @@
...
+        List<SqlParameter> parameters = statement.getParameters().stream()
+                .map(parameter -> new SqlParameter(parameter.getName().toString().toLowerCase(ENGLISH), parseTypeSignature(parameter.getType())))
+                .collect(toImmutableList());
+        TypeSignature returnType = parseTypeSignature(statement.getReturnType());
+        String description = statement.getComment().orElse("");
+        RoutineCharacteristics routineCharacteristics = new RoutineCharacteristics(
+                RoutineCharacteristics.Language.valueOf(statement.getCharacteristics().getLanguage().name()),
+                RoutineCharacteristics.Determinism.valueOf(statement.getCharacteristics().getDeterminism().name()),
+                RoutineCharacteristics.NullCallClause.valueOf(statement.getCharacteristics().getNullCallClause().name()));
+        String body = formatSql(statement.getBody(), Optional.empty());
```
Comment (rongrong) [inline presto-main/src/main/java/com/facebook/presto/execution/CreateFunctionTask.java]: It's a little strange to see `formatSql` here. Makes more sense to call it in `SHOW CREATE FUNCTION` or something like that. Consider that we didn't really do a good job in formatting expression, the original user input might even be better in a lot of cases.

Comment (caithagoras) [reply presto-main/src/main/java/com/facebook/presto/execution/CreateFunctionTask.java]: That's the only non-hacky way to convert an AST to a String representation as far as I know.

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java (hunk 21 above)
Comment (caithagoras) [reply presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java]: This is to add more flexibility to the SQL functions Presto support. I can change it to be the exact match, and properly test for the coerced case before rolling out.

**Comment by caithagoras** [comment]
Added explainDdl test.

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java (hunk 21 above)
Comment (rongrong) [reply presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java]: Let's do exact match for now. It's easier to relax constraints than add them in the future.

Comment (rongrong) [reply presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java]: You didn't use the return value btw.

Comment (caithagoras) [reply presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java]: Done

**Comment by caithagoras** [comment]
- Added tests in `TestSqlFunctions`.
- Added semantic checks for duplicating column names.
- Added TODO to ensure no SQL invoked functions in function body.
- Fix an issue for fetching function implementation in FunctionManager.

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/SemanticErrorCode.java [hunk 23]
```diff
@@ -49,6 +49,7 @@
     INVALID_LITERAL,
 
     FUNCTION_NOT_FOUND,
+    DUPLICATE_PARAMETER_DECLARATION,
```
Comment (rongrong) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/SemanticErrorCode.java]: To be consistent with `DUPLICATE_COLUMN_NAME`, maybe name this to `DUPLICATE_PARAMETER_NAME`?

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java [hunk 24]
```diff
@@ -538,6 +547,45 @@ protected Scope visitCreateView(CreateView node, Optional<Scope> scope)
             return createAndAssignScope(node, scope);
         }
 
+        @Override
+        protected Scope visitCreateFunction(CreateFunction node, Optional<Scope> scope)
+        {
+            analysis.setUpdateType("CREATE FUNCTION");
+
+            // Check parameter
+            List<String> duplicateParameters = node.getParameters().stream()
```
[the remark this answers is not in the document]
Comment (caithagoras) [reply presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java]: I need the duplicate parameter names to construct a user-friendly error message, but size checking won't tell me what are duplicated.

Alternatively, printing one name instead of a list is fine with me as well, but that does not make the code simpler.

[the remark this answers is not in the document]
Comment (rongrong) [reply presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java]: I'd just print the whole parameter list lol but sure.

**Comment by rongrong** [comment]
Thanks for working on this! Looking forward to see this in production!
